test/bench/bench.cpp

```diff
@@ -0,0 +1,242 @@
+#include "execute.hpp"
+#include "parser.hpp"
+#include <benchmark/benchmark.h>
+#include <test/utils/hex.hpp>
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+#include <memory>
+
+namespace fs = std::filesystem;
+
+namespace
+{
+constexpr auto WasmExtension = ".wasm";
+constexpr auto InputsExtension = ".inputs";
+
+inline std::string strip_space(const std::string& input)
+{
+    std::string result;
+    std::copy_if(input.begin(), input.end(), std::back_inserter(result),
+        [](auto x) noexcept { return !std::isspace(x); });
+    return result;
+}
+
+void benchmark_parse(benchmark::State& state, const fizzy::bytes& wasm_binary)
+{
+    const auto input_size = wasm_binary.size();
+    auto num_bytes_parsed = uint64_t{0};
+    for (auto _ : state)
+    {
+        const auto module = fizzy::parse(wasm_binary);
+        benchmark::DoNotOptimize(module);
+        num_bytes_parsed += input_size;
+    }
+    state.counters["size"] = benchmark::Counter(static_cast<double>(input_size));
+    state.counters["rate"] =
+        benchmark::Counter(static_cast<double>(num_bytes_parsed), benchmark::Counter::kIsRate);
+}
+
+void benchmark_instantiate(benchmark::State& state, const fizzy::bytes& wasm_binary)
+{
+    const auto module = fizzy::parse(wasm_binary);
+    for (auto _ : state)
+    {
+        auto instance = fizzy::instantiate(module, {});
+        benchmark::DoNotOptimize(instance);
+    }
+}
+
+struct ExecutionBenchmarkCase
+{
+    std::shared_ptr<const fizzy::bytes> wasm_binary;
+    std::string func_name;
+    std::vector<uint64_t> func_args;
+    fizzy::bytes memory;
+    std::optional<uint64_t> expected_result;
+    fizzy::bytes expected_memory;
+
+    void operator()(benchmark::State& state)
```

Comment:
Maybe for nice consistency this could be free function `void benchmark_execute(benchmark::State& state, const ExecutionBenchmarkCase& execution_case)` and then move `ExecutionBenchmarkCase` instance into lambda passed to `RegisterBenchmark`